### PR TITLE
add chorus fruit teleporting to existing enderpearl teleporting config option

### LINKED ISSUE #2264: No way in the config file to enable chorus fruit on claims
### Observed Behavior

Players require permission. This is a problem on PVP servers.

### Expected Behavior

Players do not require permission, when toggled in the config file.

### Reproduction steps

Walk on someone's claim. Eat a fruit.

### Stack trace or error log

_No response_

### Server version

```shell
This server is running Paper version git-Paper-462 (MC: 1.20.4) (Implementing API version 1.20.4-R0.1-SNAPSHOT) (Git: 00fd87a)
You are running the latest version
Previous version: git-Paper-454 (MC: 1.20.4)version
```

### GriefPrevention version

```shell
16.18.1
```

### Configuration

```yaml
# Default values are perfect for most servers.  If you want to customize and have a question, look for the answer here first: http://dev.bukkit.org/bukkit-plugins/grief-prevention/pages/setup-and-configuration/

GriefPrevention:
  SeaLevelOverrides:
    overworld: -1
    overworld_nether: -1
    overworld_the_end: -1
  Claims:
    Mode:
      overworld_the_end: Disabled
      overworld: Survival
      overworld_nether: Disabled
    PreventGlobalMonsterEggs: true
    PreventTheft: true
    ProtectCreatures: false
    PreventButtonsSwitches: true
    LockWoodenDoors: true
    LockTrapDoors: true
    LockFenceGates: true
    EnderPearlsRequireAccessTrust: false
    RaidTriggersRequireBuildTrust: true
    ProtectHorses: true
    ProtectDonkeys: true
    ProtectLlamas: true
    InitialBlocks: 1000
    Claim Blocks Accrued Per Hour:
      Default: 50
    Max Accrued Claim Blocks:
      Default: 100000
    Accrued Idle Threshold: -1
    AccruedIdlePercent: 0
    AbandonReturnRatio: 1.0
    AutomaticNewPlayerClaimsRadius: 4
    AutomaticNewPlayerClaimsRadiusMinimum: 0
    ExtendIntoGroundDistance: 8
    MinimumWidth: 10
    MinimumArea: 100
    MaximumDepth: -2147483648
    InvestigationTool: STICK
    ModificationTool: GOLDEN_SHOVEL
    Expiration:
      ChestClaimDays: 7
      UnusedClaimDays: 14
      AllClaims:
        DaysInactive: 30
        ExceptWhenOwnerHasTotalClaimBlocks: 16000
        ExceptWhenOwnerHasBonusClaimBlocks: 1
      AutomaticNatureRestoration:
        SurvivalWorlds: false
    AllowTrappedInAdminClaims: true
    MaximumNumberOfClaimsPerPlayer: 100
    CreationRequiresWorldGuardBuildPermission: true
    VillagerTradingRequiresPermission: true
    CommandsRequiringAccessTrust: /sethome
    DeliverManuals: false
    ManualDeliveryDelaySeconds: 30
    RavagersBreakBlocks: false
    FireSpreadsInClaims: false
    FireDamagesInClaims: false
    LecternReadingRequiresAccessTrust: true
  Spam:
    Enabled: false
    LoginCooldownSeconds: 60
    LoginLogoutNotificationsPerMinute: 5
    ChatSlashCommands: /me;/global;/local
    WhisperSlashCommands: /tell;/pm;/r;/whisper;/msg
    WarningMessage: Please reduce your noise level.  Spammers will be banned.
    BanOffenders: true
    BanMessage: Banned for spam.
    AllowedIpAddresses: 1.2.3.4; 5.6.7.8
    DeathMessageCooldownSeconds: 120
    Logout Message Delay In Seconds: 0
  PvP:
    RulesEnabledInWorld:
      overworld: true
      overworld_nether: true
      overworld_the_end: true
    ProtectFreshSpawns: false
    PunishLogout: true
    CombatTimeoutSeconds: 15
    AllowCombatItemDrop: true
    BlockedSlashCommands: '[]'
    ProtectPlayersInLandClaims:
      PlayerOwnedClaims: false
      AdministrativeClaims: false
      AdministrativeSubdivisions: false
    AllowLavaDumpingNearOtherPlayers:
      PvPWorlds: false
      NonPvPWorlds: false
    AllowFlintAndSteelNearOtherPlayers:
      PvPWorlds: false
      NonPvPWorlds: false
    ProtectPetsOutsideLandClaims: false
  Economy:
    ClaimBlocksMaxBonus: 0
    ClaimBlocksPurchaseCost: 0.0
    ClaimBlocksSellValue: 0.0
  ProtectItemsDroppedOnDeath:
    PvPWorlds: false
    NonPvPWorlds: false
  BlockLandClaimExplosions: true
  BlockSurfaceCreeperExplosions: false
  BlockSurfaceOtherExplosions: false
  LimitSkyTrees: false
  LimitTreeGrowth: false
  PistonMovement: EVERYWHERE_SIMPLE
  PistonExplosionSound: true
  FireSpreads: true
  FireDestroys: true
  AdminsGetWhispers: true
  AdminsGetSignNotifications: true
  VisualizationAntiCheatCompatMode: false
  SmartBan: false
  Mute New Players Using Banned Words: false
  MaxPlayersPerIpAddress: -1
  SilenceBans: false
  Siege:
    Worlds: []
    BreakableBlocks:
    - GRASS_BLOCK
    - DIRT
    - COBBLESTONE
    - OAK_PLANKS
    - SPRUCE_PLANKS
    - BIRCH_PLANKS
    - JUNGLE_PLANKS
    - ACACIA_PLANKS
    - DARK_OAK_PLANKS
    - SAND
    - GRAVEL
    - GLASS
    - GRASS     <-- can't understand this entry, see BukkitDev documentation
    - FERN
    - DEAD_BUSH
    - WHITE_WOOL
    - ORANGE_WOOL
    - MAGENTA_WOOL
    - LIGHT_BLUE_WOOL
    - YELLOW_WOOL
    - LIME_WOOL
    - PINK_WOOL
    - GRAY_WOOL
    - LIGHT_GRAY_WOOL
    - CYAN_WOOL
    - PURPLE_WOOL
    - BLUE_WOOL
    - BROWN_WOOL
    - GREEN_WOOL
    - RED_WOOL
    - BLACK_WOOL
    - SNOW
    - GLASS_PANE
    DoorsOpenDelayInSeconds: 300
    CooldownEndInMinutes: 60
  EndermenMoveBlocks: false
  SilverfishBreakBlocks: false
  CreaturesTrampleCrops: false
  RabbitsEatCrops: false
  HardModeZombiesBreakDoors: false
  Database:
    URL: ''
    UserName: ''
    Password: ''
  UseBanCommand: true
  BanCommandPattern: ban %name% %reason%
  Advanced:
    fixNegativeClaimblockAmounts: true
    ClaimExpirationCheckRate: 5
    OfflinePlayer_cache_days: 365
  Abridged Logs:
    Days To Keep: -1
    Included Entry Types:
      Social Activity: false
      Suspicious Activity: false
      Administrative Activity: false
      Debug: false
      Muted Chat Messages: false
  ConfigVersion: 1
```

### Plugin list

_No response_

### Running without GriefPrevention

- [ ] I attempted running the server without GriefPrevention installed.
- [ ] The problem does _not_ occur when GriefPrevention is removed from the server.

### Running with only GriefPrevention

- [ ] I attempted running only GriefPrevention on the server.
- [ ] The issue still occurs when GriefPrevention is the only plugin running.

### Running on a fresh, clean server installation

- [ ] I attempted testing for the issue on a new server.
- [ ] The issue still occurs on a new server.

### Using unmodified client

- [ ] I attempted testing for the issue with the vanilla client.
- [ ] The issue still occurs when using the vanilla client.

### We appreciate you taking the time to fill out a bug report!

- [X] I searched for similar issues before submitting this bug report.

### PR DESCRIPTION
**Pull request by QarthO**
Fixes https://github.com/GriefPrevention/GriefPrevention/issues/2264

Preventing teleporting into someone's claims with chorus fruit was already implemented in the same function as enderpearls but it wasn't being checked alongside the config ``enderPearlsRequireAccessTrust`` option. 
As a result it was always being prevented.

In the future when we plan on having api/breakage, we should change the name of the option to something else such as ``enderPearlsChorusFruitsRequireAccessTrust`` (or a better name that wont limit us in the future incase more teleporting options come available in newer version of MC.. ie ``itemTeleportRequiresAccessTrust``) but for today  I didn't want to cause any breakage with existing configs or addons. Another option is to have a new config option for it, I don't think it should be separate as they both do the same thing, and I don't see a reason to block one but not the other.

### CODE REVIEW

**Comment by Nanpure** [comment]
Why not add a separate command for pearls and chorus fruit? Someone might want one but not the other.

**Comment by Jikoo** [comment]
> Why not add a separate command for pearls and chorus fruit? Someone might want one but not the other.

> I don't think it should be separate as they both do the same thing, and I don't see a reason to block one but not the other.

That said, I do kinda agree that separate nodes makes sense. Pearls require a line of travel for the pearl, and can therefore be prevented(TM) with careful building. Chorus fruits can RNG you into the most secure vault.

**Comment by QarthO** [comment]
> Pearls require a line of travel for the pearl

This isn't true, you can glitch through floors and corners using enderpearls, (similarly to how u can glitch through the nether roof)
You can argue that you can just build a thick enough walls/roofs/ceilings, but u can do the exact same thing to prevent teleporting with chorus fruit (would need to be a bit thicker)

I'm not opposed to adding another config option though. I don't see any harm with it. This current fix was just a lot simpler

**Comment by Jikoo** [comment]
Fair point. I figured people interested in specific options like that likely have anti-glitch systems in place, but you're right, little difference from a vanilla perspective.

**Comment by RoboMWM** [comment]
I agree with qartho's reasoning. I do wonder if it's best practice tho to have the check as-is, or to refactor it to check for the config option first, then check cause?

**Comment by QarthO** [comment]
100%, I like that approach better, cleaner code, and less checks in certain situations. I can go ahead and update the PR with those changes

**Comment by QarthO** [comment]
Also looking at it, there's some siege code right below this in the teleport event, I thought siege was removed in master branch?

**Comment by RoboMWM** [comment]
> Also looking at it, there's some siege code right below this in the teleport event, I thought siege was removed in master branch?

~~That's interesting, I think I was having trouble with this file when I was reviewing the diffs... I remember looking at that area multiple times... that's weird. Siege is everywhere lol.~~ https://github.com/GriefPrevention/GriefPrevention/pull/2265#issuecomment-2026458918

**Comment by Nanpure** [comment]
In my case, it's a PVP thing. Maybe we want pearls to be active on claims but not chorus fruit.

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java [hunk 1]
```diff
@@ -1108,19 +1108,20 @@ public void onPlayerTeleport(PlayerTeleportEvent event)
 
         //FEATURE: prevent players from using ender pearls to gain access to secured claims
         TeleportCause cause = event.getCause();
-        if ((cause == TeleportCause.CHORUS_FRUIT && instance.config_claims_enderPearlsRequireAccessTrust) || (cause == TeleportCause.ENDER_PEARL && instance.config_claims_enderPearlsRequireAccessTrust))
```
Comment (RoboMWM) [inline src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: Was thinking more of `if (cause == instance.config_claims_enderPearlsRequireAccessTrust && (TeleportCause.CHORUS_FRUIT || TeleportCause.ENDER_PEARL))`

Comment (RoboMWM) [reply src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: idk if this or your first commit is more preferable though. I'm assuming performance is negligible, but if I were reading this logically to troubleshoot or such, I think this is more readable?

**Review by RoboMWM — CHANGES_REQUESTED** [review]
was thinking more of

File: src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java (hunk 1 above)
Comment (QarthO) [reply src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: Thats my first commit, I'd recommend checking my next one, which is more readable. But I'm assuming you meant

```java
if (instance.config_claims_enderPearlsRequireAccessTrust && (TeleportCause.CHORUS_FRUIT || TeleportCause.ENDER_PEARL))
```

Which my 2nd commit and your implementation has the exact same performance, no performance difference.
I do think my 2nd commit is more readable

Comment (RoboMWM) [reply src/main/java/me/ryanhamshire/GriefPrevention/PlayerEventHandler.java]: I'm not a fan of nested if's if it can be avoided

**Comment by RoboMWM** [comment]
> Also looking at it, there's some siege code right below this in the teleport event, I thought siege was removed in master branch?

must've been on an earlier commit. Just clicked the "update branch" button to pull changes in from master, and bye bye is that teleport code.

**Comment by QarthO** [comment]
There removed all of the nesting from this event
This method looks 100x more cleaner and significantly more readable

Side note: Also weird that it wasn't synced with siege removal, but looks good now

**Comment by QarthO** [comment]
Also moved some of the local variables around, so they're not taking up useless memory if the null checks aren't passed

**Review by Jikoo — APPROVED** [review]
Looks good to me. I'm assuming the previous high nesting was due to the listener having multiple features jammed into it.

**Review by RoboMWM — APPROVED** [review]

**Review by RoboMWM — APPROVED** [review]
Nice, reminds me of when I was rewriting this sort of stuff in dev/v20 a while ago c:

**Pull request merged**